modules/mqtt/source/mqtt-source.c

```diff
@@ -0,0 +1,316 @@
+                evt_tag_str("client_id", mqtt_client_options_get_client_id(&self->options)),
+                log_pipe_location_tag(&self->super.super.super.super.super));
+      return FALSE;
+    }
+
+  if (!_subscribe_topic(self))
+    return FALSE;
+
+  return TRUE;
+}
+
+static gboolean
+_unsubscribe_topic(MQTTSourceDriver *self)
+{
+  gint rc;
+  if ((rc = MQTTClient_unsubscribe(self->client, self->topic)) != MQTTCLIENT_SUCCESS)
+    {
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+_disconnect(LogThreadedFetcherDriver *s)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *)s;
+
+  _unsubscribe_topic(self);
+
+  MQTTClient_disconnect(self->client, MQTT_DISCONNECT_TIMEOUT);
+}
+
+static ThreadedFetchResult
+_receive_result_evaluation(gint rc, MQTTClient_message *message)
+{
+  if (rc == MQTTCLIENT_SUCCESS && message != NULL)
+    return THREADED_FETCH_SUCCESS;
+
+  if (rc == MQTTCLIENT_TOPICNAME_TRUNCATED && message != NULL)
+    return THREADED_FETCH_SUCCESS;
+
+  if (rc == MQTTCLIENT_SUCCESS && message == NULL)
+    // timeout
+    return THREADED_FETCH_NO_DATA;
+
+  return THREADED_FETCH_ERROR;
+}
+
+static LogThreadedFetchResult
+_fetch(LogThreadedFetcherDriver *s)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *)s;
+  ThreadedFetchResult result = THREADED_FETCH_ERROR;
+  char *topicName = NULL;
+  int topicLen;
+  MQTTClient_message *message = NULL;
+  LogMessage *msg = NULL;
+  gint rc = MQTTClient_receive(self->client, &topicName, &topicLen, &message, RECEIVE_TIMEOUT);
+
+  result = _receive_result_evaluation(rc, message);
+
+  if (result == THREADED_FETCH_SUCCESS)
+    {
+      msg = _create_message(self, (gchar *)message->payload, message->payloadlen);
+      MQTTClient_freeMessage(&message);
+      MQTTClient_free(topicName);
+    }
+
+  if (result == THREADED_FETCH_ERROR)
+    {
+      msg_error("Error while receiving msg",
+                evt_tag_str("error code", MQTTClient_strerror(rc)),
+                evt_tag_str("client_id", mqtt_client_options_get_client_id(&self->options)),
+                log_pipe_location_tag(&self->super.super.super.super.super));
+    }
+
+  return (LogThreadedFetchResult)
+  {
+    result, msg
+  };
+}
+
+static gboolean
+_init(LogPipe *s)
+{
+  MQTTSourceDriver *self = (MQTTSourceDriver *)s;
+
+  if (!self->topic)
+    {
+      msg_error("mqtt: the topic() argument is required for mqtt source",
+                log_pipe_location_tag(&self->super.super.super.super.super));
+      return FALSE;
+    }
+
+  if(!mqtt_client_options_checker(&self->options))
+    return FALSE;
+
+  if(!log_threaded_fetcher_driver_init_method(s))
+    return FALSE;
+
+  if (mqtt_client_options_get_client_id(&self->options) == NULL)
+    {
+      gchar *tmp_client_id;
+
+      tmp_client_id = g_strdup_printf("syslog-ng-%s-%s", "source", self->topic);
+
+      mqtt_client_options_set_client_id(&self->options, tmp_client_id);
+      g_free(tmp_client_id);
```

Comment:
Can you move this to a separate function? Thanks!
Also, you can replace the first `%s` with `source` directly.

Reply:
commit: 6bd600370ba3f4cb03e096be0b01c522f0ce43be